### PR TITLE
Calendar sync (ICS plus Google/Outlook)

### LINKED ISSUE #333: Calendar sync (ICS plus Google/Outlook)
## Summary
Integrate calendar sync from ICS and popular providers (Google/Outlook).

## Scope
- Add calendar source configuration and fetch/refresh pipeline.
- Map calendar events into scheduling guidance and planning surfaces.
- Handle timezone, recurrence, and provider-specific edge cases.

## Acceptance criteria
- Calendar events from supported sources influence focus planning reliably.
- Sync behavior remains correct across timezone and recurrence scenarios.

### PR DESCRIPTION
**Pull request by utilForever**
## What

Adds calendar busy-window sync from ICS feeds (including Google and Outlook ICS URLs), integrates the cached busy windows into schedule messaging, and introduces a `--calendar-sync` CLI workflow for manual refresh plus diagnostics output.

## Why

Issue #333 requested calendar-aware planning so users can see upcoming conflicts and active busy periods directly in focustime scheduling surfaces. This change adds a dedicated calendar domain (separate from encrypted backup sync), keeps recurrence/timezone handling deterministic, and exposes an explicit refresh path through CLI.

Closes #333.

### Required

- [x] `cargo check --all` passes
- [x] `cargo fmt --all -- --check` passes
- [x] `cargo clippy --all-targets -- -D warnings` passes
- [x] `cargo test --all` passes
- [x] I linked the related issue (for example: `Closes #25`)

### Functional Validation

- [x] Pomodoro timer behavior was verified for this change (if applicable) (N/A: no timer state-machine behavior changes)
- [x] Site blocking behavior was verified for this change (if applicable) (N/A: no blocking backend changes)
- [x] WakaTime integration behavior was verified for this change (if applicable) (N/A: no WakaTime behavior changes)
- [x] I added or updated tests for changed behavior (if applicable)

### Configuration & Docs

- [x] User-facing docs were updated (`README.md` or relevant docs, if applicable)
- [x] New dependencies/configuration are documented (if applicable)
- [x] No sensitive values or credentials were introduced

### If Applicable

- [ ] Security impact considered (run `cargo audit` locally if needed)
- [x] Breaking behavior changes are clearly described in this PR (none)

## Summary by CodeRabbit

* **New Features**
  * Calendar sync from ICS sources with persisted cache and schedule warnings for busy/overlap periods
* **CLI**
  * Added --calendar-sync command with text/JSON output; errors redact sensitive URL/details
* **Configuration**
  * New calendar_sync settings: enabled flag, refresh interval, lookahead window, and normalized/deduped source list with default naming
* **Documentation**
  * Updated architecture and README with calendar module diagram, config guide, and CLI examples
* **Tests**
  * Added unit tests for parsing, recurrence expansion, normalization, and CLI behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/utilForever/focustime/pull/372?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/calendar.rs [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+                    end_epoch_secs: end.timestamp(),
+                });
+            }
+        }
+        let Some(next) = date.succ_opt() else {
+            break;
+        };
+        date = next;
+    }
+}
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/calendar.rs:420]: Implemented in db77e75. `expand_event` no longer breaks when an advanced date equals `end_date`; it now only breaks when `succ_opt()` returns `None`. I also added regression test `parse_ics_includes_weekly_occurrence_on_range_end_date` to cover this boundary.

File: src/cli/execute.rs [hunk 2]
```diff
@@ -2034,6 +2036,24 @@ fn execute_sync_restore_command(
...
+    let config = AppConfig::load().normalized();
+    let result = crate::calendar::sync_from_config(&config.calendar_sync, chrono::Local::now())?;
+    let payload = CalendarSyncCommandOutput {
+        action: "calendar-sync",
+        synced_at_epoch_secs: result.synced_at_epoch_secs,
+        source_count: result.source_count,
+        windows_count: result.windows.len(),
+        error_count: result.source_errors.len(),
+        errors: result.source_errors,
+    };
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/cli/execute.rs:2054]: Implemented in 22153b6. Calendar sync CLI error output now redacts URL query and fragment data through `redact_calendar_sync_error`, and I added focused tests for redaction behavior and non-URL token preservation.

File: src/config/tests.rs [hunk 3]
```diff
@@ -207,6 +207,7 @@ fn round_trip_full_config() {
             time_step_minutes: 20,
             delay_secs: 15 * 60,
         },
+        calendar_sync: CalendarSyncConfig::default(),
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/config/tests.rs:210]: Implemented in ce701bb (with formatting in 9dd1796). Added direct `CalendarSyncConfig` normalization tests covering refresh/lookahead clamping, source deduplication, and auto-generated names for blank source names.

File: src/cli/execute.rs [hunk 4]
```diff
@@ -2034,6 +2036,84 @@ fn execute_sync_restore_command(
...
+        || token.starts_with("http://")
+        || token.starts_with("https://")
+}
+
+fn redact_url_like_token(token: &str) -> String {
+    let without_query = token.split_once('?').map_or(token, |(base, _)| base);
+    without_query
+        .split_once('#')
+        .map_or(without_query, |(base, _)| base)
+        .to_string()
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/cli/execute.rs]: Implemented in 096d7ac. I updated URL redaction to preserve only scheme and host while redacting path/query/fragment (`https://host/<redacted>` when a path is present), and added coverage for path redaction plus host-only URL handling.

File: src/cli/execute.rs [hunk 5]
```diff
@@ -2034,6 +2036,92 @@ fn execute_sync_restore_command(
...
+        .map_or(without_fragment, |(base, _)| base);
+    let Some((scheme, remainder)) = without_query.split_once("://") else {
+        return without_query.to_string();
+    };
+    let host = remainder.split('/').next().unwrap_or(remainder);
+    if remainder.contains('/') {
+        format!("{scheme}://{host}/<redacted>")
+    } else {
+        format!("{scheme}://{host}")
+    }
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/cli/execute.rs:2118]: Implemented in e5a1f1c. `redact_url_like_token` now strips URL userinfo before composing redacted output, so credentials like `user:pass@` are never emitted. Added regression test `redact_calendar_sync_error_redacts_url_userinfo`.
